### ISSUE TITLE
Does DBDish for mysql error - Unsupported Concurrency: Connection used by multiple threads simultaneously 

### ISSUE THREAD

**Issue by melezhik** [raku-community-modules/DBIish#222]
Hi! I am having this in my cro app application that reuse the same db handler across http requests:

```
📓 sparky job-queue-fs: create trigger file: /home/sph/.sparky/projects/raku.notify/.triggers/lcxejwqavbhizpoktsym.1518045
⚠ sparky DBDish::mysql: Unsupported Concurrency: Connection used by multiple threads simultaneously (-1)
⚠ sparky   in method protect-connection at /home/sph/.raku/sources/D8985539A1D1B912F6B444D889F20895BF534A23 (DBDish::Connection) line 119
⚠ sparky   in method protect-connection at /home/sph/.raku/sources/D8985539A1D1B912F6B444D889F20895BF534A23 (DBDish::Connection) line 119
⚠ sparky   in method prepare at /home/sph/projects/sparky/lib/.precomp/A7BDC13A4A1A8FB19924ED5DD1A71063344AC17E/C4/C47588D2D9336749CA96D9909E8ED48607FB927B line 1
⚠ sparky   in block  at bin/sparky-web.raku line 371
⚠ sparky   in block  at /home/sph/.raku/sources/3A6026658FBAAD4D3FE8D607027D323A500AB5C3 (Cro::HTTP::Router) line 136
⚠ sparky   in block  at /home/sph/.raku/sources/3A6026658FBAAD4D3FE8D607027D323A500AB5C3 (Cro::HTTP::Router) line 132
⚠ sparky 
⚠ sparky double free or corruption (!prev)
```

**Comment by jnthn** [comment]
In general, one cannot reuse database connection objects across threads. This isn't anything Raku-specific; the same is observed over different database drivers and bindings to them in different programming languages. The typical solution is to use a connection pool; I'm not sure if this module offers one (and that's why I've tended to use `DB::Pg` instead in Cro apps, since it has had one out of the box from the start).

**Comment by rbt** [comment]
@melezhik

Try DBIish::Pool. Grab one connection for each request, and discard it at the end of the request.

@jnthn

> I'm not sure if this module offers one (and that's why I've tended to use DB::Pg instead in Cro apps, since it has had one out of the box from the start).

I originally created DBIish::Pool with intention of merging into this driver directly via #171, but there was a preference to have it as a separate module.

**Comment by melezhik** [comment]
Hi @rbt - thanks - I will try DBIish::Pool. and let you guys, know .

Thanks

Alexey


чт, 31 мар. 2022 г. в 06:01, rbt ***@***.***>:

**Comment by melezhik** [comment]
@jnthn , yeah DB::Pg is one of the options here, though I am hesitant to switch to `postgresql` from  `mysql` which is on my view is easier to maintain ....

**Comment by jnthn** [comment]
> yeah DB::Pg is one of the options here, though I am hesitant to switch to postgresql from mysql which is on my view is easier to maintain ....

D'oh, yes, that wan't so useful an answer on my part. :-) What I should have added is that there is [https://modules.raku.org/dist/DB::MySQL:cpan:CTILMES](DB::MySQL) too, which is pretty much the `DB::Pg` approach but for MySQL. That said, if you're already using `DBIish`, the pool module is probably the path of least resistance from where you are.

**Comment by melezhik** [comment]
@jnthn ++ thanks for that, @patrickbkr  already mentioned `DB::MySQL` in raku irc channel, I will keep it as a second possible solution ...

**Comment by CurtTilmes** [comment]
`DB::Pg` is in pretty good shape. `DB::MySQL` has some fundamental brokenness I haven't been able to fix.
